File: 2017/05/Day5.py
def execute_pt1(offsets: list) -> int:
    pos = 0
    moves = 0
    while pos >= 0 and pos < len(offsets):
        new = offsets[pos]
        offsets[pos] += 1
        pos += new
        moves += 1
    return moves


def execute_pt2(offsets: list) -> int:
    pos = 0
    moves = 0
    while pos >= 0 and pos < len(offsets):
        new = offsets[pos]
        if new >= 3:
            offsets[pos] -= 1
        else:
            offsets[pos] += 1
        pos += new
        moves += 1
    return moves


input_data = list(map(int, open("input.txt").read().split("\n")))

print(execute_pt1(input_data[:]))
print(execute_pt2(input_data[:]))

File: 2017/05/Day5.fs
module AdventOfCode.Solutions.Y2017.D05

open System.IO

[<TailCall>]
let rec jump
    (offsets: int array)
    (position: int)
    (steps: int)
    (calculateNewOffset: int -> int)
    : int =
    if position < 0 || position >= offsets.Length then
        steps
    else
        let positionOffset = offsets[position]

        jump
            (offsets
             |> Array.updateAt position (calculateNewOffset positionOffset))
            (position + positionOffset)
            (steps + 1)
            calculateNewOffset

let inputData = File.ReadAllLines "input.txt" |> Array.map int

let solve () =
    jump inputData 0 0 (fun offset -> offset + 1) |> printfn "%d"

    jump inputData 0 0 (fun offset ->
        if offset >= 3 then offset - 1 else offset + 1)
    |> printfn "%d"
